[managed-ledger] Do not send duplicate reads to BK/offloaders

eolivelli opened the pull request:
### Motivation
When you have many subscriptions on the same topic and they are catching up (for instance after some downtime of the consumer application) Pulsar will start to perform many reads from BK for the same entries.

### Modifications
Prevent concurrent reads of the same entries.

### Verifying this change

This change is already covered by existing tests.

I also tested  manually that with this fix we are reducing a lot the pressure on the bookies. This is something we cannot code in a integration test as it will use too many resources on CI.

With some tests with only 64 subscription we saw a reduction from 100K reads/s from BK to 250 reads/s.

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +271,77 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
 
+    @AllArgsConstructor
+    private static final class ReadEntriesCallbackWithContext {
+        ReadEntriesCallback callback;
+        Object ctx;
```
make fields `final`?

eolivelli replied:
sure. fixed

lhotari reviewed (approved):
LGTM, Good work @eolivelli

nicoloboschi reviewed (approved):
+1

codelipenghui reviewed (commented):
I understand the problem that the PR wants to resolve for now. And I think it's a good point of the reads improvement.

I have some questions

1. We are using the PendingReadKey as the key to determining whether the new one is duplicated read or not. It looks can only cover the case that all the subscriptions with the same read position. It will be great if we can have a more general solution for the duplicated reads which able to handle the case like [0, 20], [2, 22]
2. After this change, even if the subscription with different read position e.g. sub-a 1:10, sub-b 3:20. We will always introduce the new heap memory overhead.

I haven't thought of a specific way yet, but I think it's very necessary to discuss looking for a general solution for duplicated reads.

I think we can start the discussion on the mailing list. So that many contributors can share their ideas.

eolivelli commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -52,9 +58,18 @@ public class RangeEntryCacheImpl implements EntryCache {
+    private static class PendingReadKey {
+        private final long ledgerId;
+        private final long startEntry;
+        private final long endEntry;
```
done

zymap reviewed (approved):
Great work!

zymap commented:
> 1. We are using the PendingReadKey as the key to determining whether the new one is duplicated read or not. It looks can only cover the case that all the subscriptions with the same read position. It will be great if we can have a more general solution for the duplicated reads which able to handle the case like [0, 20], [2, 22]

I have an idea about this that is to split the read range into small pieces, then different read can share the pieces to construct their read. There will have a tradeoff, if the pieces are smaller, the different reads will share more. 
Something like this:
![image](https://user-images.githubusercontent.com/24502569/186559152-9b486e00-9254-4aac-8562-0b11e2e3a411.png)

Jason918 reviewed (approved):
LGTM

eolivelli commented:
@zymap thank you for your idea.
I am also thinking to the same think in order to allow more chances to match the same range.
There is an open discussion on the ML, maybe you can post your idea there.

eolivelli commented:
@codelipenghui @Jason918 @zymap @lhotari @nicoloboschi 
I have updated this patch.

We did more testing and actually trying to find pending reads with ranges that "include" the requested range allows us to have many more "hits".

![image](https://user-images.githubusercontent.com/9469110/186709693-83cb6dd2-4619-4c48-b112-2223e0089bf8.png)

in the picture:
- green line is for "perfect matches" (previous version of the patch)
- the blue line is the rate of "partial matches" (excluding the perfect matches)
- the yellow line is "no pending read to attach to"

the rate of "partial matches" is usually higher or equals to the rate of "misses" both during tailing reads and catch up reads.

eolivelli commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +276,132 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
+            if (existing != null) {
+                return existing;
+            }
+            // TODO: use a sorted map
```
I am not 100% that we need a SortedMap to make the search more efficient.
sorting the map by the "startEntry" does not help much.

in my tests it is more likely that the "endEntry" is in common (because during tailing reads you hit the wall of the last written message)

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -295,51 +439,70 @@ private void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boo
+
+            boolean listenerAdded = false;
+            while (!listenerAdded) {
+                AtomicBoolean createdByThisThread = new AtomicBoolean();
+                CachedPendingRead cachedPendingRead = findBestCandidate(key,
```
I think we can only make the findBestCandidate() return null if no candidate and then create a new one to continue the read operation.

So that we don't need `AtomicBoolean createdByThisThread = new AtomicBoolean();` and the code is easy to read.

eolivelli replied:
findBestCandidate is kind of a computeIfAbsent method (when I was not looking for "includes" it was actually computeIfAbsent on a ConcurrentHashMap).
So I have to create the object and put it into the map inside the "lock"

codelipenghui replied:
Yes, I just want to avoid creating `AtomicBoolean` here.
Can we use the callback size of the `cachedPendingRead` or a boolean inside `CachedPendingRead`?

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +276,130 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
+            return result;
+        }
+
+        public void attach(CompletableFuture<List<EntryImpl>> handle) {
```
Do we need to remove the `CachedPendingRead` from the `cachedPendingReads`?

eolivelli replied:
it is done here
https://github.com/apache/pulsar/pull/17241/files#diff-c55509e3ab1389d89a58fd564f2e318dbb95f50121ab33c729a7ca4a21d02ef1R340

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -295,51 +439,70 @@ private void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boo
     public void clear() {
         Pair<Integer, Long> removedPair = entries.clear();
         manager.entriesRemoved(removedPair.getRight(), removedPair.getLeft());
+        cachedPendingReads.clear();
```
I noticed the element of the `cachedPendingReads` only be removed here. This one will be called if all the cached data should be clean up

It's better to add a unit test for this one to make sure we will not introduce any heap memory leak.

eolivelli replied:
the elements here are one entry per ledger.
the per ledger maps are evicted here
https://github.com/apache/pulsar/pull/17241/files#diff-c55509e3ab1389d89a58fd564f2e318dbb95f50121ab33c729a7ca4a21d02ef1R340

We could remove the entry for a ledger in case of rollover.
but I thought it requires more coordination.
I will try to improve this

codelipenghui replied:
> We could remove the entry for a ledger in case of rollover.
but I thought it requires more coordination.
I will try to improve this

Yes, we are on the same page.

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +276,130 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
+
+    private CachedPendingRead findBestCandidate(PendingReadKey key, Map<PendingReadKey, CachedPendingRead> ledgerCache,
+                                                AtomicBoolean created) {
+        synchronized (ledgerCache) {
```
We can move the synchronized to method

eolivelli replied:
I want to synchronize only on the Map that is for the specific "ledger", not on the whole RangeEntryCacheImpl
this way the lock is more fine grained

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +276,130 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
+                        }
+                    } else {
+                        for (ReadEntriesCallbackWithContext callback : callbacks) {
+                            List<EntryImpl> copy = new ArrayList<>(entriesToReturn.size());
+                            long callbackStartEntry = callback.startEntry;
+                            long callbackEndEntry = callback.endEntry;
```
Looks like we can change to 

long callbackStartEntry = callback.startEntry;
long callbackEndEntry = callback.endEntry
List<EntryImpl> copy = new ArrayList<>(callbackEndEntry - callbackStartEntry + 1);

eolivelli replied:
good point

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -256,6 +276,130 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
+            }, ml.getExecutor().chooseThread(ml.getName())).exceptionally(exception -> {
+                synchronized (CachedPendingRead.this) {
+                    for (ReadEntriesCallbackWithContext callback : callbacks) {
+                        if (exception instanceof BKException
```
We should unwrap the exception

codelipenghui replied:
And looks like we can just use `createManagedLedgerException` directly without the `if` check here.

```java
    public static ManagedLedgerException createManagedLedgerException(Throwable t) {
        if (t instanceof org.apache.bookkeeper.client.api.BKException) {
            return createManagedLedgerException(((org.apache.bookkeeper.client.api.BKException) t).getCode());
        } else if (t instanceof CompletionException
                && !(t.getCause() instanceof CompletionException) /* check to avoid stackoverlflow */) {
            return createManagedLedgerException(t.getCause());
        } else {
            log.error("Unknown exception for ManagedLedgerException.", t);
            return new ManagedLedgerException("Other exception", t);
        }
    }
```
It already handled the `CompletionException` and `TooManyRequestsException`

eolivelli replied:
sure, thanks. previous code had a "invalidateLedger" that has been moved

eolivelli commented:
@zymap @codelipenghui @Jason918 @lhotari 
I have added one last step to the patch and now it is really ready:
- added the ability to reuse pending reads that partially overlap with the requested range
- refactored the new code to a dedicated class, in order to reduce the complexity of the code (and to ease testability)

I tried to do something more sophisticated, like reusing only overlapping pending reads only if the amount of overlapping entries is big enough, but in my testing I have seen that it seems always a good idea to reuse pending reads.

Jason918 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java:
```diff
@@ -0,0 +1,443 @@
+                    } else if (reminderOnRight != null && reminderOnLeft == null) {
+                        foundButMissingSomethingOnRight = new FindPendingReadOutcome(entry.getValue(),
+                                null, reminderOnRight);
+                    }
```
we can `break` the iteration if overlaps

eolivelli replied:
I pushed a small improvement in this direction. good idea. we may save some cycles

Jason918 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java:
```diff
@@ -0,0 +1,443 @@
+                        for (ReadEntriesCallbackWithContext callback : callbacks) {
+                            long callbackStartEntry = callback.startEntry;
+                            long callbackEndEntry = callback.endEntry;
+                            List<EntryImpl> copy = new ArrayList<>((int) (callbackEndEntry - callbackEndEntry + 1));
```
```suggestion
                            List<EntryImpl> copy = new ArrayList<>((int) (callbackEndEntry - callbackStartEntry + 1));
```

eolivelli replied:
good catch

eolivelli commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java:
```diff
@@ -0,0 +1,443 @@
+        final PendingReadKey key = new PendingReadKey(firstEntry, lastEntry);
+
+        Map<PendingReadKey, PendingRead> pendingReadsForLedger =
+                cachedPendingReads.computeIfAbsent(lh.getId(), (l) -> new ConcurrentHashMap<>());
```
we can invalidate it in `invalidateAllEntries` that is called when the ledger is trimmed.

fixed. good idea (also suggested by @codelipenghui )

eolivelli commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:
```diff
@@ -295,51 +299,61 @@ private void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boo
+        }
+    }
+
+    CompletableFuture readFromStorage(ReadHandle lh, long firstEntry, long lastEntry, boolean shouldCacheEntry) {
```
fixed

eolivelli commented:
@Jason918 thanks for your suggestions. I have applied them

zymap reviewed (approved):
Nice improve! Looks good to me.

Do we need to add a unit test for the PendingReadsManager?

Jason918 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java:
```diff
@@ -0,0 +1,450 @@
+                        // consider the other options
+                        break;
+                    } else if (reminderOnLeft != null && reminderOnRight == null) {
+                        foundButMissingSomethingOnLeft = new FindPendingReadOutcome(entry.getValue(),
```
Sorry, I don't get why we don't break here? Any reason `right` is better than `left`?
From what I see, there are some aspects to consider :
1. Less `rangeEntryCache.asyncReadEntry0` calls is better. So `left` and `right` > `both`.
2. Less cycles in this for loop. So just break if we find any matched result.
3. Less entries to read from BK. Which we are not considering here. This needs more cycles or other complex data structures.

eolivelli replied:
below we have to choose what happens in case we have:
- foundButMissingSomethingOnLeft
- foundButMissingSomethingOnRight
- foundButMissingSomethingOnBoth

Currently we give preference to foundButMissingSomethingOnRight. So we can break the loop when we found it.

We can add in the future something more sophisticated to decide "the best" operation to reuse.

I did a few tests and  "missing on right" is the thing that happens more frequently, both on tailing reads and when I have many subscriptions that are catching up (even if I give preference to the other options).
This makes sense as the subscriptions are moving forward so when a new read comes and there is already one pending it is common that the new read misses the new entries "on the right" (on the tail of the topic).

After re-reading the code If we break here when we find "anything" we could lose the chance to find the "match include" case, that is the absolutely the best, as we don't need further reads.
So I am going to remove this "break"

Jason918 replied:
> So I am going to remove this "break"


If so, since we are going through all the `PendingReadKey` in ledgerCache. I seems we can simple achieve the goal of "Less entries to read from BK" by comparing the number of entries to read. This can also solve the case that there are multiply `foundButMissingSomethingOnLeft` or `foundButMissingSomethingOnRight`.

eolivelli commented:
@zymap 
> Do we need to add a unit test for the PendingReadsManager?
I haven't pushed this part of my patch. I will update this PR soon. Thanks for the reminder.

eolivelli commented:
@zymap tests added

michaeljmarshall commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java:
```diff
@@ -0,0 +1,447 @@
+
+    private static final Counter COUNT_ENTRIES_NOTREAD_FROM_BK = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_entries_notread")
+            .help("Total number of entries not read from BK")
+            .register();
+
+    private static final Counter COUNT_PENDING_READS_MATCHED = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_matched")
+            .help("Pending reads reused with perfect range match")
+            .register();
+    private static final Counter COUNT_PENDING_READS_MATCHED_INCLUDED = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_matched_included")
+            .help("Pending reads reused by attaching to a read with a larger range")
+            .register();
+    private static final Counter COUNT_PENDING_READS_MISSED = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_missed")
+            .help("Pending reads that didn't find a match")
+            .register();
+
+    private static final Counter COUNT_PENDING_READS_MATCHED_OVERLAPPING_MISS_LEFT = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_matched_overlapping_miss_left")
+            .help("Pending reads that didn't find a match but they partially overlap with another read")
+            .register();
+
+    private static final Counter COUNT_PENDING_READS_MATCHED_BUT_OVERLAPPING_MISS_RIGHT = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_matched_overlapping_miss_right")
+            .help("Pending reads that didn't find a match but they partially overlap with another read")
+            .register();
+
+    private static final Counter COUNT_PENDING_READS_MATCHED_BUT_OVERLAPPING_MISS_BOTH = Counter
+            .build()
+            .name("pulsar_ml_cache_pendingreads_matched_overlapping_miss_both")
+            .help("Pending reads that didn't find a match but they partially overlap with another read")
+            .register();
```
I know it is a little late, but I am wondering if there is a reason we created so many metrics here instead of using a single metric with different labels to differentiate certain outcomes. These metrics could probably have been 3 metrics each with 2 or 3 unique labels. For example, `COUNT_PENDING_READS_MATCHED_BUT_OVERLAPPING_MISS_RIGHT` could be `COUNT_PENDING_READS_MATCHED_BUT_OVERLAPPING` with a label indicating that the miss was `right`, `left`, or `both`. In my experience, this makes made Grafana dashboards simpler because you can aggregate the whole metric or break it out by label. I am not aware of any performance implications.

Here is the prometheus guide to metric and label names that supports using labels to differentiate a single metric: https://prometheus.io/docs/practices/naming/#labels.
